src/tqec/sketchup/block_graph.py

```diff
@@ -551,14 +580,117 @@ def display(
         pipe_length: float = 2.0,
     ) -> ColladaDisplayHelper:
         """Display the block graph in 3D."""
+        bytes_buffer = BytesIO()
         from tqec.sketchup.collada import (
             display_collada_model,
             write_block_graph_to_dae_file,
         )
 
-        bytes_buffer = BytesIO()
         write_block_graph_to_dae_file(self, bytes_buffer, pipe_length)
         return display_collada_model(
             filepath_or_bytes=bytes_buffer.getvalue(),
             write_html_filepath=write_html_filepath,
         )
+
+    def to_circuit(self, dimension: LinearFunction) -> cirq.Circuit:
+        """Build and return the quantum circuit representing the computation.
+
+        A block graph can be interpreted as a quantum computation.
+        Each cube represents a self contained layer of computations.
+        Each pipe modifies the initial and final layer of the cubes it connects.
+
+        Raises:
+            TQECException: if any of the circuits obtained by instantiating the
+                computational blocks is contains a qubit that is not a cirq.GridQubit.
+
+        Returns:
+            a cirq.Circuit instance representing the full computation.
+        """
+        blocks: dict[Position3D, StandardComputationBlock] = {}
+        for cube in self.cubes:
+            blocks[cube.position] = cube_to_block(cube, dimension)
+        for pipe in self.pipes:
+            u_computation, v_computation = (
+                blocks[pipe.u.position],
+                blocks[pipe.v.position],
+            )
+            u_computation.replace_boundary_plaquettes(pipe.direction)
+            v_computation.replace_boundary_plaquettes(pipe.direction, outgoing=False)
+
+        instantiated_scheduled_blocks: list[ScheduledCircuit] = []
+        depth = 0
+        for position, block in blocks.items():
+            spatially_shifted_circuit = block.instantiate().transform_qubits(
+                partial(_shift_qubits, position)
+            )
```

Comment:
I think this is broken, there are two different coordinate systems here:

1. the coordinate system used to place cubes and pipes,
2. the physical qubits coordinate system.

Here `position` is in the first coordinate system, but `_shift_qubits` is expecting the second coordinate system. I think, each entry in position `position` should be multiplied by the number of qubits needed by each block spatial dimension (`2k + 2`? I might be wrong on the exact number.)

Reply:
I didn't touch this anymore. First, I wanted to get the computation graph right. Once I have the graph, I can go back to fixing this